prov/rxm/src/rxm_ep.c

```diff
@@ -635,6 +637,144 @@ static ssize_t rxm_rma_iov_init(struct rxm_ep *rxm_ep, void *buf,
 	return sizeof(*rma_iov) + sizeof(*rma_iov->iov) * count;
 }
 
+static inline ssize_t
+rxm_ep_transmit_generic_lightweight(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn,
+				    size_t len, uint64_t data, uint64_t flags,
+				    uint64_t tag, uint8_t op, struct rxm_tx_buf **tx_buf)
+{
+	*tx_buf = RXM_TX_BUF_GET(rxm_ep);
+	if (OFI_UNLIKELY(!*tx_buf)) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_DATA, "TX queue full!\n");
+		return -FI_EAGAIN;
+	}
+
+	(*tx_buf)->hdr.msg_ep = rxm_conn->msg_ep;
+
+	(*tx_buf)->pkt.ctrl_hdr.version = OFI_CTRL_VERSION;
+	(*tx_buf)->pkt.ctrl_hdr.conn_id = rxm_conn->handle.remote_key;
+	(*tx_buf)->pkt.ctrl_hdr.type = ofi_ctrl_data;
+	(*tx_buf)->pkt.hdr.version = OFI_OP_VERSION;
+	(*tx_buf)->pkt.hdr.op = op;
+	(*tx_buf)->pkt.hdr.size = len;
+	(*tx_buf)->pkt.hdr.tag = tag;
+	rxm_op_hdr_process_flags(&(*tx_buf)->pkt.hdr, flags, data);
+
+	return FI_SUCCESS;
+}
+
+static inline ssize_t
+rxm_ep_transmit_generic(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn, void *context,
+			uint8_t count, size_t len, uint64_t data, uint64_t flags,
+			uint64_t tag, uint8_t op, uint64_t comp_flags,
+			struct rxm_tx_buf **tx_buf, struct rxm_tx_entry **tx_entry)
+{
+	ssize_t ret;
+
+	*tx_buf = RXM_TX_BUF_GET(rxm_ep);
+	if (OFI_UNLIKELY(!*tx_buf)) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_DATA, "TX queue full!\n");
+		ret = -FI_EAGAIN;
+		goto err1;
+	}
+
+	*tx_entry = rxm_tx_entry_get(&rxm_ep->send_queue);
+	if (OFI_UNLIKELY(!*tx_entry)) {
+		ret = -FI_EAGAIN;
+		goto err2;
+	}
+
+	(*tx_entry)->ep = rxm_ep;
+	(*tx_entry)->count = count;
+	(*tx_entry)->context = context;
+	(*tx_entry)->flags = flags;
+	(*tx_entry)->tx_buf = *tx_buf;
+	(*tx_entry)->comp_flags |= comp_flags | FI_SEND;
+
+	(*tx_buf)->hdr.msg_ep = rxm_conn->msg_ep;
+
+	(*tx_buf)->pkt.ctrl_hdr.version = OFI_CTRL_VERSION;
+	(*tx_buf)->pkt.ctrl_hdr.conn_id = rxm_conn->handle.remote_key;
+	(*tx_buf)->pkt.hdr.version = OFI_OP_VERSION;
+	(*tx_buf)->pkt.hdr.op = op;
+	(*tx_buf)->pkt.hdr.size = len;
+	(*tx_buf)->pkt.hdr.tag = tag;
+	rxm_op_hdr_process_flags(&(*tx_buf)->pkt.hdr, flags, data);
```

Comment:
Can rxm_ep_transmit_generic_lightweight be called here instead? (as long as it doesn't increase the # of instructions).

Reply:
Sure, nice point!
This change decreases the # of instruction on 1. It seems that this is due to we fill `tx_buf` immediately after allocating memory

Reply:
Attaching an output from the instructions counting tool for the 256 bytes buffer:
- [OFI/rxm new](https://github.com/ofiwg/libfabric/files/1625203/ofi_rxm_256_new.txt)
- [OFI/rxm old](https://github.com/ofiwg/libfabric/files/1625204/ofi_rxm_256_old.txt)